include/cuco/detail/probe_sequences.cuh

```diff
@@ -0,0 +1,165 @@
+  __device__ const_iterator get_slots() const noexcept { return slots_; }
+
+ protected:
+  iterator slots_;
+  const std::size_t capacity_;
+};  // class probe_sequence_base
+
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          typename Hash1           = cuco::detail::MurmurHash3_32<Key>,
+          typename Hash2           = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class double_hashing : public probe_sequence_base<Key, Value, CGSize, Scope> {
+ public:
+  using iterator = typename probe_sequence_base<Key, Value, CGSize, Scope>::iterator;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::capacity_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::slots_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::cg_size;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::vector_width;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::uses_vector_load;
+
+  __host__ __device__ explicit double_hashing(iterator slots, std::size_t capacity) noexcept
+    : probe_sequence_base<Key, Value, CGSize, Scope>{slots, capacity}
+  {
+  }
+
+  template <typename CG>
+  __device__ iterator initial_slot(CG const& g, Key const k) noexcept
+  {
+    std::size_t index;
+    auto const hash_value = hash1_(k);
+    if constexpr (uses_vector_load()) {
+      step_size_ = (hash2_(k + 1) % (capacity_ / (cg_size() * vector_width()) - 1) + 1) *
+                   cg_size() * vector_width();
+      index = hash_value % (capacity_ / (cg_size() * vector_width())) * cg_size() * vector_width() +
+              g.thread_rank() * vector_width();
+    } else {
+      step_size_ = (hash2_(k + 1) % (capacity_ / cg_size() - 1) + 1) * cg_size();
+      index      = (hash_value + g.thread_rank()) % capacity_;
+    }
+    return slots_ + index;
+  }
+
+  __device__ iterator next_slot(iterator s) noexcept
+  {
+    std::size_t index = s - slots_;
+    return &slots_[(index + step_size_) % capacity_];
+  }
+
+ private:
+  std::size_t step_size_;
+  Hash1 hash1_{};
+  Hash2 hash2_{};
+};  // class double_hashing
+
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          typename Hash            = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class linear_probing : public probe_sequence_base<Key, Value, CGSize, Scope> {
+ public:
+  using iterator = typename probe_sequence_base<Key, Value, CGSize, Scope>::iterator;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::capacity_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::slots_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::cg_size;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::vector_width;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::uses_vector_load;
+
+  __host__ __device__ explicit linear_probing(iterator slots, std::size_t capacity)
```

Comment:
Same thing here, this should accept a `Hash` object that is defaulted to being default constructured.